grpcinterceptor: Use ctxutil to detect context cancellation

miretskiy opened the pull request:
Previous commit #76306 arranged for interceptors to run for all rpcs (local or otherwise).  In doing so, it added an additional go routine for each RPC whose purpose was to detect context completion, and to arrange for the cleanup function to run when that happened.

`ctxutil.WhenDone` (added by #103866/commits/dfe4ef82810aaf81ae21afecce83cbc1fb6783dd) allows for the same functionality without the additional goroutine cost.

Informs #107053

Release note: None

Commit:
grpcinterceptor: Use ctxutil to detect context cancellation

Previous commit #76306 arranged for interceptors to run
for all rpcs (local or otherwise).  In doing so, it added
an additional goroutine for each RPC whose purpose was to
detect context completion, and to arrange for the cleanup
function to run when that happened.

`ctxutil.WhenDone` (added by #103866/commits/dfe4ef82810aaf81ae21afecce83cbc1fb6783dd)
allows for the same functionality without the additional
goroutine cost.

Informs #107053

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107079)
<!-- Reviewable:end -->

knz commented:
We have two problems here.
@yuzefovich as the reviewer of #103866 I'd like you to double check what I write below.

### 1. Goroutine saving - are you sure?

> allows for the same functionality without the additional goroutine cost.

This is not true in general. Have you opened the black box? Look at [the go source code, this `propagateCancel` spawns a goroutine to do the job](https://github.com/golang/go/blob/master/src/context/context.go#L504):

   ```go
                atomic.AddInt32(&goroutines, +1)
                go func() {
                        select {
                        case <-parent.Done():
                                child.cancel(false, parent.Err())
                        case <-child.Done():
                        }
                }()
   ```

### 2. Possible goroutine leak!

> ctxutil.WhenDone (added by https://github.com/cockroachdb/cockroach/pull/103866/commits/dfe4ef82810aaf81ae21afecce83cbc1fb6783dd) 

I double checked this and I don't like what I am seeing. 

If you look at the go source code for this internal propagateCancel" function, you see that it spawns a goroutine. That goroutine **only ever terminates if either the child or parent is canceled**.

So _what if they don't get canceled at all?_

At the very least with our code base, when someone spawns a goroutine they know they have to listen on the stopper ShouldQuiesce channel. This is also encouraged in the API of our `retry` package.

This new `ctxutil.WhenDone` does not encourage that. Worse yet, _it terminates immediately_ (i.e. it does not block execution in the caller) which means there is no synchronization on termination of that implicit goroutine.

## Recommendation: remove ctxutil altogether

If the above is confirmed, I would recommend removing the current impl of `ctxutil` altogether.

If we really need this functionality, **we need to build a patch to the go runtime to do what we want** (we have a patch infrastructure now! let's use it!) while ensuring that stray goroutines don't leak.

erikgrinaker commented:
Thanks for looking into this @knz, but I think your concerns can be assuaged. `WhenDone()` can only take a context with `Done() != nil`, i.e. a cancellable context derived from `WithCancel` or `WithTimeout`:

https://github.com/cockroachdb/cockroach/blob/dfe4ef82810aaf81ae21afecce83cbc1fb6783dd/pkg/util/ctxutil/context.go#L34-L36

>If you look at the go source code for this internal propagateCancel" function, you see that it spawns a goroutine.

Only if the parent context isn't cancellable. We've already established that it must be, so it takes this code path here, which doesn't spawn a goroutine:

https://github.com/golang/go/blob/5fe3f0a265c90a9c0346403742c6cafeb154503b/src/context/context.go#L473-L487

>So what if they don't get canceled at all? Like, if our code paths pass a context.Background or similar?

We've established that the context must be cancellable, so the caller is responsible for avoiding goroutine leaks by calling the associated cancel function.

I think the only concern here would be if we use a custom context implementation that overrides `Done()`, in which case `parentCancelCtx` won't be able to resolve the parent `cancelCtx` and attach to it.

That said, I'm not super-thrilled about accessing Go internals like this. This stuff is internal for a reason.

I'll give this a closer review.

erikgrinaker commented:
FWIW, I also verified that this gets rid of the superfluous goroutine.

miretskiy commented:
@knz -- what @erikgrinaker said.  

> Have you opened the black box? 
Obviously, I have in order to implement ctxutil package.  If you do some shenanigans and override context, then
the default code path is taken that does spin up goroutine.  But, we don't do that; and in the common case of 
context.WithTimeout/WithCancel, the code does not spawn anything.

> If we really need this functionality, we need to build a patch to the go runtime to do what we want (we have a patch infrastructure now! let's use it!) while ensuring that stray goroutines don't leak.

Goroutines do not leak with this change.  The behavior is no different from the default implementation of context
which spins up go routine if the implementation is anything other than the 2 "approved" cancelers.
The last time context.go changed was the first time go was released.  I think a ~20 year lifetime is not that terrible, and of course, we could try patching things locally.

re @erikgrinaker 
> That said, I'm not super-thrilled about accessing Go internals like this. This stuff is internal for a reason.
Me neither;   Having said that, a) this particular implementation does not change frequently -- first time in 1.20; b) we do occasionally access those internal implementations ("git grep linkname") -- it's not done frequently, nor in a wonton way.
If there was another way to do it, I would.

I have searched quite extensively on this topic; there were proposals to extend context.go somehow; there was always a pushback.

miretskiy commented:
> We have two problems here.
> @yuzefovich as the reviewer of #103866 I'd like you to double check what I write below.

@knz -- I don't think there are two problems here.  
The reason why the change reviewed by @yuzefovich was backed out is that while
this implementation *is* faster than checking for ->Done() (~8x), it is not 128 (or even 1024) times faster since
that's how frequent the Done checks are in vectorized engine.

erikgrinaker commented on pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:
```diff
@@ -338,28 +337,25 @@ func newTracingClientStream(
+		// [1] https://pkg.go.dev/google.golang.org/grpc#ClientConn.NewStream
+		finishFunc(nil /* err */)
+	}); err != nil && !errors.Is(err, ctxutil.ErrNeverCompletes) {
+		panic(err)
```
Let's propagate the error. I know this can't currently happen, but still, I don't like latent panics and the only caller has an error return path.

miretskiy replied:
Logic reworked.

erikgrinaker reviewed (approved):
This seems fine, and gets the job done.

Somewhat uneasy about `ctxutil.WhenDone()`, especially considering it's otherwise unused. Can we add in some `CrdbTestBuild` assertions ensuring that it always does what it's supposed to, either here or separately? We'll also want to let this bake for a while before backporting.

knz commented:
I am friendly to the change now. You have successfully changed my opinion.

miretskiy commented:
@erikgrinaker @knz 
Please take another look.  I expanded test coverage in ctxutil package; Added utility function to check
if we can directly propagate context, and added a check which triggers panic under test.

erikgrinaker commented on pkg/util/ctxutil/context.go:
```diff
@@ -24,19 +23,39 @@ type WhenDoneFunc func(err error)
-// If the context never becomes done, returns ErrNeverCompletes error.
-func WhenDone(parent context.Context, done WhenDoneFunc) error {
+// parent context becomes done and returns true.
+// If the context never becomes done, returns false.
```
```suggestion
// If the context is non-cancellable (i.e. `Done() == nil`), returns false and never calls the function.
```

erikgrinaker commented on pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:
```diff
@@ -338,28 +339,36 @@ func newTracingClientStream(
+		// standard context.Context.
+		// But, be safe and loudly fail tests in case somebody introduces strange
+		// context implementation.
+		if buildutil.CrdbTestBuild && !ctxutil.CanDirectlyDetectCancellation(ctx) {
```
Let's move this into `WhenDone()`, I think we always want to assert this. If we ever make a custom context implementation that doesn't derive from the parent's cancellation, we can come up with something else then. Better safe than sorry.

erikgrinaker commented on pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:
```diff
@@ -338,28 +339,36 @@ func newTracingClientStream(
+		// But, be safe and loudly fail tests in case somebody introduces strange
+		// context implementation.
+		if buildutil.CrdbTestBuild && !ctxutil.CanDirectlyDetectCancellation(ctx) {
+			panic(fmt.Sprintf(
```
```suggestion
			log.Fatalf(ctx,
```

miretskiy commented:
bors r+
